File: 0074.Search_a_2D_Matrix.js
//0ms

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if(matrix==null||matrix.length==0||matrix[0].length==0){
        return false;
    }
    let m= matrix.length;
    let n= matrix[0].length;
    let low=0;
    let high=m*n-1;
    while(low<=high){
        let mid=low+Math.floor((high-low)/2);
        let midvalue=matrix[Math.floor(mid/n)][Math.floor(mid%n)];
        if(midvalue==target){
                return true;
        }else if(midvalue<target){
                low=mid+1;
        }else{
                high=mid-1;
        }
    }
    return false;
};

File: 0074.Search_a_2D_Matrix.cpp
//0ms

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        int l=0;
        int r = m*n-1;

        while(l<=r){
            int mid = l+(r-l)/2;
            int rows=mid/n;
            int col=mid%n;

            if(matrix[rows][col]==target){
                return true;
            }
            else if(matrix[rows][col]<target){
                l=mid+1;
            }
            else{
                r=mid-1;
            }
        }
        return false;
    }
};
